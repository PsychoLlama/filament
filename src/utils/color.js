import invariant from 'invariant';
import Color from 'hue-colors';

// Used when a light or a group of lights is powered off.
export const BLACK = '#000000';

// Represents the type of color encoding last used.
export const colorTypes = {
  TEMP: 'ct',
  HSB: 'hs',
  XY: 'xy',
};

// A map of color encodings to their hex code converters.
export const colorExtractors = {
  // Hue, saturation, brightness.
  [colorTypes.HSB]: ({ hue, sat, bri }) => {
    invariant(typeof hue === 'number', 'Missing hue value');
    invariant(typeof sat === 'number', 'Missing saturation value');
    invariant(typeof bri === 'number', 'Missing brightness value');

    return Color.fromHsb(hue, sat, bri);
  },

  // XY color coordinates.
  [colorTypes.XY]: ({ xy, bri }) => {
    invariant(typeof bri === 'number', 'Missing brightness value');
    invariant(Array.isArray(xy), 'Missing XY value');

    return Color.fromCIE(xy[0], xy[1], bri);
  },

  // Color temperature (orange-ish to white).
  [colorTypes.TEMP]: ({ ct, bri }) => {
    invariant(typeof bri === 'number', 'Missing brightness value');
    invariant(typeof ct === 'number', 'Missing temperature value');

    return Color.fromCt(ct, bri);
  },
};

/**
 * Takes light state and turns it into a CSS color code.
 * @param  {Object} state - State of a light or group of lights.
 * @param  {String} state.colormode - Either `xy`, `hs`, or `ct`.
 * @param  {Number} [state.hue] - Number between 0-0xffff.
 * @param  {Number} [state.sat] - Saturation.
 * @param  {Number} [state.ct] - Color temperature.
 * @param  {Array<Number>} [state.xy] - XY color space coordinates.
 * @param  {Number} state.bri - Brightness.
 * @return {String} - CSS hex color code, hash character and all.
 */
export default state => {
  // Default to HSB just in case Philips adds another color mode.
  const extractor =
    colorExtractors[state.colormode] || colorExtractors[colorTypes.HSB];

  // Bless you, hue-colors. Bless you.
  const hexColorCode = extractor(state).toHex();

  return `#${hexColorCode}`;
};
