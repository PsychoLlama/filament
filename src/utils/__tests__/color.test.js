import color, { colorTypes } from '../color';

describe('The color converter', () => {
  it('passes sanity checks', () => {
    const result = color({ hue: 10, sat: 20, bri: 30 });

    expect(result).toEqual(expect.any(String));
    expect(result[0]).toBe('#');
    expect(result.length).toBe(7);
  });

  it('defaults to HSB', () => {
    const result = color({ hue: 10, sat: 20, bri: 30 });

    expect(result).toEqual(expect.any(String));
    expect(result).toMatchSnapshot();
  });

  it('understands XY color coordinates', () => {
    const result = color({ xy: [0.2, 0.7], bri: 10, colormode: colorTypes.XY });

    expect(result).toMatchSnapshot();
  });

  it('understands temperature encoding', () => {
    const result = color({ colormode: colorTypes.TEMP, ct: 238, bri: 32 });

    expect(result).toMatchSnapshot();
  });

  it('throws if HSB values are omitted', () => {
    expect(() => color({ sat: 20, bri: 30 })).toThrow(/hue/i);
    expect(() => color({ hue: 10, bri: 30 })).toThrow(/sat/i);
    expect(() => color({ hue: 10, sat: 20 })).toThrow(/bri/i);
  });

  it('throws if CT values are omitted', () => {
    const colormode = colorTypes.TEMP;

    expect(() => color({ colormode, ct: 100 })).toThrow(/bri/);
    expect(() => color({ colormode, bri: 100 })).toThrow(/temperature/);
  });

  it('throws if XY values are omitted', () => {
    const colormode = colorTypes.XY;

    expect(() => color({ colormode, xy: [0.2, 0.6] })).toThrow(/bri/);
    expect(() => color({ colormode, bri: 250 })).toThrow(/xy/i);
  });
});
