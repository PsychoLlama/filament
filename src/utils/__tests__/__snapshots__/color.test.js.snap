// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The color converter defaults to HSB 1`] = `"#1e1c1c"`;

exports[`The color converter understands XY color coordinates 1`] = `"#38380a"`;

exports[`The color converter understands temperature encoding 1`] = `"#201b20"`;
